From e4ff052d8d14b277593697c127df4ca75b73fb4b Mon Sep 17 00:00:00 2001
From: Bo Fang <quic_bofang@quicinc.com>
Date: Thu, 25 Aug 2022 00:42:05 +0800
Subject: [PATCH 2/2] waylandsink: support ubwc modifier and
 zwp_linux_dmabuf_v1 version3

To support UBWC, must upgrade zwp_linux_dmabuf_v1 from ver1 to ver3. Then,
dmabuf protocol could contain modifier information. UBWC and other information,
like security, could be transmitted from wayland client to wayland server
through this modifier channel.

To interact with upstream plugin, introduced "compression" field in caps. This
"compression" could be "linear" or "ubwc", based on wayland server reported
modifier capability.

Added NV12_10LE32 UBWC format support, it also named as Q12A or TP10. HW
decoder won't output NV12_10LE32 linear.

Upstream-Status: Pending

Change-Id: I09bcca6202def252bd0f7e580d2d170d8a5c4b5d
Signed-off-by: Bo Fang <quic_bofang@quicinc.com>
Signed-off-by: Lily Li <quic_lali@quicinc.com>
---
 .../ext/wayland/gstwaylandsink.c              | 31 ++++++++++++++++++-
 .../ext/wayland/gstwaylandsink.h              |  1 +
 .../gst-plugins-bad/ext/wayland/wldisplay.c   | 22 ++++++++++++-
 .../gst-plugins-bad/ext/wayland/wldisplay.h   |  1 +
 .../ext/wayland/wllinuxdmabuf.c               |  9 +++++-
 .../ext/wayland/wlvideoformat.c               |  1 +
 6 files changed, 62 insertions(+), 3 deletions(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 7a3c1f54b1..a3074b5e15 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -62,6 +62,7 @@ enum
 enum
 {
   PROP_0,
+  PROP_DISABLE_UBWC,
   PROP_DISPLAY,
   PROP_FULLSCREEN
 };
@@ -72,7 +73,7 @@ GST_DEBUG_CATEGORY (gstwayland_debug);
 #define WL_VIDEO_FORMATS \
     "{ BGRx, BGRA, RGBx, xBGR, xRGB, RGBA, ABGR, ARGB, RGB, BGR, " \
     "RGB16, BGR16, YUY2, YVYU, UYVY, AYUV, NV12, NV21, NV16, NV61, " \
-    "YUV9, YVU9, Y41B, I420, YV12, Y42B, v308, P010_10LE }"
+    "YUV9, YVU9, Y41B, I420, YV12, Y42B, v308, P010_10LE, NV12_10LE32 }"
 
 static GstStaticPadTemplate sink_template = GST_STATIC_PAD_TEMPLATE ("sink",
     GST_PAD_SINK,
@@ -207,6 +208,11 @@ gst_wayland_sink_class_init (GstWaylandSinkClass * klass)
           "display name to connect to, if not supplied via the GstContext",
           NULL, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
 
+  g_object_class_install_property (gobject_class, PROP_DISABLE_UBWC,
+      g_param_spec_boolean ("disable-ubwc", "Disable UBWC",
+          "Disable UBWC in the video sink",
+          FALSE, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+
   g_object_class_install_property (gobject_class, PROP_FULLSCREEN,
       g_param_spec_boolean ("fullscreen", "Fullscreen",
           "Whether the surface should be made fullscreen ", FALSE,
@@ -246,6 +252,11 @@ gst_wayland_sink_get_property (GObject * object,
       g_value_set_string (value, sink->display_name);
       GST_OBJECT_UNLOCK (sink);
       break;
+    case PROP_DISABLE_UBWC:
+      GST_OBJECT_LOCK (sink);
+      g_value_set_boolean (value, sink->disable_ubwc);
+      GST_OBJECT_UNLOCK (sink);
+      break;
     case PROP_FULLSCREEN:
       GST_OBJECT_LOCK (sink);
       g_value_set_boolean (value, sink->fullscreen);
@@ -269,6 +280,11 @@ gst_wayland_sink_set_property (GObject * object,
       sink->display_name = g_value_dup_string (value);
       GST_OBJECT_UNLOCK (sink);
       break;
+    case PROP_DISABLE_UBWC:
+      GST_OBJECT_LOCK (sink);
+      sink->disable_ubwc = g_value_get_boolean (value);
+      GST_OBJECT_UNLOCK (sink);
+      break;
     case PROP_FULLSCREEN:
       GST_OBJECT_LOCK (sink);
       gst_wayland_sink_set_fullscreen (sink, g_value_get_boolean (value));
@@ -482,6 +498,10 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
     gint i;
     guint fmt;
     GstVideoFormat gfmt;
+    GValue compression_val = { 0, };
+    GValue compressions = { 0, };
+    g_value_init (&compression_val, G_TYPE_STRING);
+    g_value_init (&compressions, GST_TYPE_LIST);
 
     g_value_init (&shm_list, GST_TYPE_LIST);
     g_value_init (&dmabuf_list, GST_TYPE_LIST);
@@ -516,6 +536,15 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
     gst_structure_take_value (gst_caps_get_structure (caps, 1), "format",
         &dmabuf_list);
 
+    //add compression in caps
+    g_value_set_static_string (&compression_val, "linear");
+    gst_value_list_append_value (&compressions, &compression_val);
+    if ((sink->display->server_modifier_caps & 0x3) && !sink->disable_ubwc )  {
+      g_value_set_static_string (&compression_val, "ubwc");
+      gst_value_list_append_value (&compressions, &compression_val);
+    }
+    gst_structure_take_value (gst_caps_get_structure (caps, 1), "compression", &compressions);
+
     GST_DEBUG_OBJECT (sink, "display caps: %" GST_PTR_FORMAT, caps);
   }
 
diff --git a/ext/wayland/gstwaylandsink.h b/ext/wayland/gstwaylandsink.h
index 7aabb6ff0e..82a59aa6b7 100644
--- a/ext/wayland/gstwaylandsink.h
+++ b/ext/wayland/gstwaylandsink.h
@@ -63,6 +63,7 @@ struct _GstWaylandSink
   gboolean fullscreen;
 
   gchar *display_name;
+  gboolean disable_ubwc;
 
   gboolean redraw_pending;
   GMutex render_lock;
diff --git a/ext/wayland/wldisplay.c b/ext/wayland/wldisplay.c
index f326091990..93a308abf6 100644
--- a/ext/wayland/wldisplay.c
+++ b/ext/wayland/wldisplay.c
@@ -147,8 +147,28 @@ dmabuf_format (void *data, struct zwp_linux_dmabuf_v1 *zwp_linux_dmabuf,
     g_array_append_val (self->dmabuf_formats, format);
 }
 
+static void
+dmabuf_modifier(void *data, struct zwp_linux_dmabuf_v1 *zwp_linux_dmabuf_v1, uint32_t format, uint32_t modifier_hi, uint32_t modifier_lo)
+{
+  GstWlDisplay *self = data;
+  guint64 modifier = ((uint64_t) modifier_hi << 32) | modifier_lo;
+  GST_INFO ("wayland server report format 0x%x(%"GST_FOURCC_FORMAT") modifier 0x%llx", format, GST_FOURCC_ARGS(format), modifier);
+  if (modifier == DRM_FORMAT_MOD_QCOM_COMPRESSED) {
+    if (format == DRM_FORMAT_NV12) {
+      self->server_modifier_caps |= 1;
+    }else if (format == GST_MAKE_FOURCC('Q','1','2','A')) {
+      self->server_modifier_caps |= 2;
+    }
+  }
+
+  if (gst_wl_dmabuf_format_to_video_format (format) != GST_VIDEO_FORMAT_UNKNOWN) {
+    g_array_append_val (self->dmabuf_formats, format);
+  }
+}
+
 static const struct zwp_linux_dmabuf_v1_listener dmabuf_listener = {
   dmabuf_format,
+  dmabuf_modifier,
 };
 
 gboolean
@@ -235,7 +255,7 @@ registry_handle_global (void *data, struct wl_registry *registry,
         wl_registry_bind (registry, id, &wp_viewporter_interface, 1);
   } else if (g_strcmp0 (interface, "zwp_linux_dmabuf_v1") == 0) {
     self->dmabuf =
-        wl_registry_bind (registry, id, &zwp_linux_dmabuf_v1_interface, 1);
+        wl_registry_bind (registry, id, &zwp_linux_dmabuf_v1_interface, version >= 3 ? 3 : 1);
     zwp_linux_dmabuf_v1_add_listener (self->dmabuf, &dmabuf_listener, self);
   }
 }
diff --git a/ext/wayland/wldisplay.h b/ext/wayland/wldisplay.h
index f2025a6a03..13b9e87738 100644
--- a/ext/wayland/wldisplay.h
+++ b/ext/wayland/wldisplay.h
@@ -62,6 +62,7 @@ struct _GstWlDisplay
   struct zwp_linux_dmabuf_v1 *dmabuf;
   GArray *shm_formats;
   GArray *dmabuf_formats;
+  guint64 server_modifier_caps;
 
   /* private */
   gboolean own_display;
diff --git a/ext/wayland/wllinuxdmabuf.c b/ext/wayland/wllinuxdmabuf.c
index 96487d142a..ceda7ec30a 100644
--- a/ext/wayland/wllinuxdmabuf.c
+++ b/ext/wayland/wllinuxdmabuf.c
@@ -31,6 +31,8 @@
 GST_DEBUG_CATEGORY_EXTERN (gstwayland_debug);
 #define GST_CAT_DEFAULT gstwayland_debug
 
+static G_DEFINE_QUARK (FBufModifierQuark, gst_fbuf_modifier_qdata);
+
 typedef struct
 {
   GMutex lock;
@@ -79,10 +81,15 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
   struct zwp_linux_buffer_params_v1 *params;
   gint64 timeout;
   ConstructBufferData data;
+  guint64 modifier, *p_modifier;
 
   g_return_val_if_fail (gst_wl_display_check_format_for_dmabuf (display,
           GST_VIDEO_INFO_FORMAT (info)), NULL);
 
+  p_modifier = buf ? gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (buf), gst_fbuf_modifier_qdata_quark()) : NULL;
+  modifier = p_modifier ? *p_modifier : DRM_FORMAT_MOD_INVALID;
+  GST_INFO_OBJECT (display, "Found modifier: gstbuf %p, %p, modifier val 0x%llx", buf, p_modifier, modifier);
+
   mem = gst_buffer_peek_memory (buf, 0);
   format = gst_video_format_to_wl_dmabuf_format (GST_VIDEO_INFO_FORMAT (info));
 
@@ -111,7 +118,7 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
       GstMemory *m = gst_buffer_peek_memory (buf, mem_idx);
       gint fd = gst_dmabuf_memory_get_fd (m);
       zwp_linux_buffer_params_v1_add (params, fd, i, m->offset + skip,
-          stride, 0, 0);
+          stride, modifier >> 32, modifier & 0xffffffff);
     } else {
       GST_ERROR_OBJECT (mem->allocator, "memory does not seem to contain "
           "enough data for the specified format");
diff --git a/ext/wayland/wlvideoformat.c b/ext/wayland/wlvideoformat.c
index dd5d801d8f..1ea1ee21cc 100644
--- a/ext/wayland/wlvideoformat.c
+++ b/ext/wayland/wlvideoformat.c
@@ -67,6 +67,7 @@ static const wl_VideoFormat wl_formats[] = {
   {WL_SHM_FORMAT_YUV422, DRM_FORMAT_YUV422, GST_VIDEO_FORMAT_Y42B},
   {WL_SHM_FORMAT_YUV444, DRM_FORMAT_YUV444, GST_VIDEO_FORMAT_v308},
   {WL_SHM_FORMAT_P010, DRM_FORMAT_P010, GST_VIDEO_FORMAT_P010_10LE},
+  {0xffffffff, GST_MAKE_FOURCC('Q','1','2','A'), GST_VIDEO_FORMAT_NV12_10LE32},
 };
 
 enum wl_shm_format
-- 
2.34.1

